src/storage/src/source/generator/upsert.rs

```diff
@@ -0,0 +1,556 @@
+    fn finished(&self) -> bool {
+        self.round >= self.snapshot_rounds
+    }
+
+    /// Produce a batch of message into `buffer`, of size `batch_size`. Advances the current
+    /// batch and round counter.
+    fn produce_batch(
+        &mut self,
+        buffer: &mut Vec<(
+            (usize, Result<SourceMessage, SourceReaderError>),
+            MzOffset,
+            Diff,
+        )>,
+        value_buffer: &mut Vec<u8>,
+    ) -> u64 {
+        if self.finished() {
+            return 0;
+        }
+
+        let rng = self.rng.get_or_insert_with(|| {
+            // Consistently seeded with the source see, the round (offset), and the
+            // partition.
+            let mut seed = [0; 32];
+            seed[0..8].copy_from_slice(&self.seed.to_le_bytes());
+            seed[8..16].copy_from_slice(&self.round.to_le_bytes());
+            seed[16..24].copy_from_slice(&self.pi.partition.to_le_bytes());
+            StdRng::from_seed(seed)
+        });
+
+        let partition = self.pi.partition;
+        buffer.extend(self.pi.take(usize::cast_from(self.batch_size)).map(|key| {
+            rng.fill_bytes(value_buffer.as_mut_slice());
+            let msg = (
+                0,
+                Ok(SourceMessage {
+                    key: Row::pack_slice(&[Datum::UInt64(key)]),
+                    value: Row::pack_slice(&[Datum::UInt64(partition), Datum::Bytes(value_buffer)]),
+                    metadata: Row::default(),
+                }),
+            );
+            (msg, MzOffset::from(self.round), 1)
+        }));
+
+        self.produced_batches += 1;
+
+        if self.produced_batches == self.expected_batches {
+            self.round += 1;
+            self.produced_batches = 0;
+        }
+        self.batch_size
+    }
+}
+
+/// A struct that produces batches of data for the post-snapshotting phase.
+struct UpdateProducer {
+    // The key iterator for the partition.
+    pi: PartitionKeyIterator,
+    // The batch size.
+    batch_size: u64,
+    // The next offset to produce updates at.
+    next_offset: u64,
+    // The source-level seed for the rng.
+    seed: u64,
+    // The number of offsets we expect to be part of the `quick_rounds`.
+    expected_quick_offsets: u64,
+}
+
+impl UpdateProducer {
+    fn new(
+        partition: u64,
+        partitions: u64,
+        keys: u64,
+        snapshot_rounds: u64,
+        quick_rounds: u64,
+        batch_size: u64,
+        seed: u64,
+        next_offset: u64,
+    ) -> Self {
+        // The non-snapshot offset * the number of keys in a batch partitioned across the key
+        // space is the _key_ of the next batch at `next_offset`.
+        let start_key =
+            (((next_offset - snapshot_rounds) * batch_size * partitions) + partition) % keys;
+
+        // We expect to emit 1 batch per offset.
+        let expected_quick_offsets =
+            ((keys / partitions / batch_size) * quick_rounds) + snapshot_rounds;
+
```

Comment:
this is a little hard to wrap my head around, and I wonder if it's because `round` and `offset` are being used for the same concept? Is that correct or am I misunderstanding?

Reply:
They are very very similar, I will clarify!